src/freenet/client/filter/TheoraPacketFilter.java

```diff
@@ -1,148 +1,287 @@
+									QRSIZES[qti][pli][qri] = input.readInt(ilog(62 - qi)) + 1;
+
+									qi = qi + QRSIZES[qti][pli][qri];
+									qri++;
+
+									QRBMIS[qti][pli][qri] = input.readInt(ilog(NBMS - 1));
+
+									if (qi < 63)
+										continue;
+									else if (qi > 63)
+										throw new UnknownContentTypeException("qi = " + qi + "; qi > 63 - The stream is undecodable.");
+
+									break;
+								}
+
+								NQRS[qti][pli] = qri;
+							}
+						}
+					}
+
+					int[][] HTS = new int[80][0];
+					for (int hti = 0; hti < 80; hti++)
+						HTS[hti] = readHuffmanTable("", HTS[hti], input);
+
+					if (logMINOR)
+						Logger.minor(this, "SETUP_HEADER: left " + input.available() + " bytes (should be 0)");
+
+					expectedPacket = Packet.FRAME;
 					break;
+
+				case FRAME:
 			}
 		} catch(IOException e) {
-			if(logMINOR) Logger.minor(this, "In theora parser caught "+e, e);
+			if (logMINOR) Logger.minor(this, "In Theora parser caught " + e, e);
 			throw e;
 		}
 
 		return packet;
 	}
 
-	private long decode32bitIntegerFrom8BitChunks(DataInputStream input) throws IOException {
-		int LEN0 = input.readUnsignedByte();
-		int LEN1 = input.readUnsignedByte();
-		int LEN2 = input.readUnsignedByte();
-		int LEN3 = input.readUnsignedByte();
-		return LEN0|(LEN1 << 8)|(LEN2 << 16)|(LEN3 << 24);
-	}
+	private void checkMagicHeader(byte[] typeAndMagicHeader, byte expectedType) throws IOException {
+		if (logMINOR) Logger.minor(this, "Header type: " + typeAndMagicHeader[0]);
+
+		if (typeAndMagicHeader[0] != expectedType)
+			throw new UnknownContentTypeException("Header type: " + typeAndMagicHeader[0] + ", expected: " + expectedType);
 
-	private void checkMagicHeader(byte[] typeAndMagicHeader) throws IOException {
-		for(int i=0; i < magicNumber.length; i++) {
-			if(typeAndMagicHeader[i+1] != magicNumber[i])
+		for (int i=0; i < magicNumber.length; i++) {
+			if (typeAndMagicHeader[i+1] != magicNumber[i])
 				throw new UnknownContentTypeException("Packet header magicNumber[" + i + "]: " + typeAndMagicHeader[i+1]);
 		}
 	}
+
+	// The minimum number of bits required to store a positive integer `a` in
+	// two’s complement notation, or 0 for a non-positive integer a.
+	private int ilog(int a) {
+		if (a <= 0)
+			return 0;
+
+		int n = 0;
+		while (a > 0) {
+			a >>= 1;
+			n++;
+		}
+		return n;
+	}
+
+	private int[] readHuffmanTable(String HBITS, int[] HTS, BitInputStream input) throws IOException {
+		if (HBITS.length() > 32)
+			throw new UnknownContentTypeException("HBITS = " + HBITS +
+					"; HBITS is longer than 32 bits in length - The stream is undecodable.");
+
+		int ISLEAF = input.readBit();
+		if (ISLEAF == 1) {
+			if (HTS.length == 32)
+				throw new UnknownContentTypeException("HTS[hti] = " + Arrays.toString(HTS) +
+						"; HTS[hti] is already 32 - The stream is undecodable.");
+			int TOKEN = input.readInt(5);
+
+			HTS = Arrays.copyOf(HTS, HTS.length + 1);
+			HTS[HTS.length - 1] = TOKEN;
+		} else {
+			HBITS += 0;
```

Comment:
This is the same as `HBITS += "0"` but more confusing. (Also, still not a big fan of modifying method parameters.)

Reply:
On second thought, does this have to be a String at all? It looks like you’re just using it for counting, the actual value doesn’t seem to matter, except for logging…